### PR TITLE
bot: auto-merge pull request under conditions and do not run ci when no code changes

### PR DESCRIPTION
**Pull request by leseb**
**Description of your changes:**

We now allow mergify to auto-merge pull requests against the master
branch under the following conditions:

* all CI tests passed
* the PR has no label 'do-not-merge'
* the PR has the label 'ready-to-merge'
* has been approved by at least one maintainer
* has not requested changes

Also, the CI will not run when no Go files have been modified.
We just validate Rook builds correctly.

Signed-off-by: Sébastien Han <seb@redhat.com>

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

[skip ci]

### CODE REVIEW

File: .github/workflows/canary-integration-test.yml
Comment (subhamkrai) [inline .github/workflows/canary-integration-test.yml:5]: Do we want to run the integration test if there is changes in `.py` or `.sh` file?

File: .mergify.yml
Comment (subhamkrai) [inline .mergify.yml:61]: Jenkins test needed if no code changes or only docs changes?

File: .mergify.yml
Comment (subhamkrai) [inline .mergify.yml:28]: doubts: 
1) approved reviewers after approving need to add label `ready-to-merge` to get it merged?
2) not sure if any of these conditions ``` "#review-requested=0", "#changes-requested-reviews-by=0", "#commented-reviews-by=0```" needed.

File: .mergify.yml
Comment (leseb) [reply .mergify.yml:61]: For now, Jenkins will go green anyway since it detects it is a doc change..

File: .mergify.yml
Comment (leseb) [reply .mergify.yml:28]: > doubts:
> 
> 1. approved reviewers after approving need to add label `ready-to-merge` to get it merged?

Yes. So basically, you are done reviewing, you approve, you think twice, then set the label, go do something else, and once CI is green it will get merged. This extra step to set the label is more of a security.

> 2. not sure if any of these conditions ` "#review-requested=0", "#changes-requested-reviews-by=0", "#commented-reviews-by=0`" needed.

I don't think we need those.

File: .github/workflows/canary-integration-test.yml
Comment (leseb) [reply .github/workflows/canary-integration-test.yml:5]: Most likely yes.

File: .mergify.yml
Comment (subhamkrai) [reply .mergify.yml:28]: Thanks for explaining.

File: .github/workflows/build.yml
Comment (travisn) [inline .github/workflows/build.yml:20]: Helm doesn't complain if I build without the VERSION set. When you're building locally do you have to set it as well?

File: .github/workflows/canary-integration-test.yml
Comment (travisn) [inline .github/workflows/canary-integration-test.yml:5]: There may be some PRs that we want to run the full tests, even though these files didn't change, for example if we update the base operator image. Could we also trigger these actions with a label when we want to override the file extensions?

File: .github/workflows/build.yml
Comment (travisn) [inline .github/workflows/build.yml:4]: How about a comment to explain the paths-ignore?

File: .mergify.yml
Comment (travisn) [inline .mergify.yml:43]: Every time we add a new action we will need to add it to this list. Is there no way to have it check for overall success instead?

File: .mergify.yml
Comment (travisn) [inline .mergify.yml:40]: If the action is skipped (no .go files), will the action pass this `check-success` validation?

File: .mergify.yml
Comment (travisn) [inline .mergify.yml:51]: Oh I see now, there is a separate rule to merge if there was no code change

**Review by travisn — CHANGES_REQUESTED** [review]

File: .mergify.yml
Comment (leseb) [reply .mergify.yml:43]: Not possible and explicitly unrecommended https://docs.mergify.io/conditions.html#validating-all-status-check

File: .mergify.yml
Comment (leseb) [reply .mergify.yml:40]: Handled by rule on L51

File: .github/workflows/canary-integration-test.yml
Comment (leseb) [reply .github/workflows/canary-integration-test.yml:5]: I don't think it's possible :-/

File: .github/workflows/build.yml
Comment (leseb) [reply .github/workflows/build.yml:20]: It seems to be happening here only...

File: .github/workflows/canary-integration-test.yml
Comment (leseb) [reply .github/workflows/canary-integration-test.yml:5]: Not possible...

Comment (travisn) [reply .github/workflows/canary-integration-test.yml:5]: Ok if we can't control it with a label, then what if we add more paths to trigger the tests, such as `build/*`, `tests/*`, and `images/*`? I'd rather run the tests too often than miss sometimes when needed.

Comment (leseb) [reply .github/workflows/canary-integration-test.yml:5]: sounds good

Comment (leseb) [reply .github/workflows/canary-integration-test.yml:5]: Build on image changes?

Comment (travisn) [reply .github/workflows/canary-integration-test.yml:5]: changes under `images` are rare, but yes I think it makes sense to run integration tests in that case

**Review by travisn — APPROVED** [review]

**Pull request merged**
